lib/host/wasi/inode-linux.cpp

```diff
@@ -769,14 +769,39 @@ WasiExpect<Poller> INode::pollOneoff(__wasi_size_t NSubscriptions) noexcept {
   }
 }
 
-WasiExpect<void> INode::getAddrinfo(const char *NodeStr, const char *ServiceStr,
+WasiExpect<void> INode::getAddrinfo(std::string_view Node,
+                                    std::string_view Service,
                                     const __wasi_addrinfo_t &Hint,
                                     uint32_t MaxResLength,
                                     Span<__wasi_addrinfo_t *> WasiAddrinfoArray,
                                     Span<__wasi_sockaddr_t *> WasiSockaddrArray,
                                     Span<char *> AiAddrSaDataArray,
                                     Span<char *> AiCanonnameArray,
                                     /*Out*/ __wasi_size_t &ResLength) noexcept {
+  std::string NodeStr;
+  const char *NodeCStr = nullptr;
+  if (!Node.empty()) {
+    if (const auto pos = Node.find_first_of('\0');
+        pos != std::string_view::npos) {
+      NodeCStr = Node.data();
+    } else {
+      NodeStr = Node;
+      NodeCStr = NodeStr.c_str();
+    }
+  }
+
+  std::string ServiceStr;
+  const char *ServiceCStr = nullptr;
+  if (!Service.empty()) {
+    if (const auto pos = Service.find_first_of('\0');
+        pos != std::string_view::npos) {
+      ServiceCStr = Service.data();
+    } else {
+      ServiceStr = Service;
+      ServiceCStr = ServiceStr.c_str();
+    }
+  }
+
```

Comment:
Looks like these two transformations are the same. Should we have a helper function to do this?

Reply:
changed to helper function